src/test/java/com/google/sps/MapServletTest.java

```diff
@@ -0,0 +1,186 @@
+
+    // create tripDay entity
+    Entity tripDayEntity = new Entity(TripDay.QUERY_STRING, tripEntity.getKey());
+    tripDayEntity.setProperty("origin", INPUT_DESTINATION);
+    tripDayEntity.setProperty("destination", INPUT_DESTINATION);
+    tripDayEntity.setProperty("date", TRIP_DAY_OF_TRAVEL);
+    datastore.put(tripDayEntity);
+
+    // create location entities and put it into datastore
+    Entity DomeEntity = new Entity(TripDay.LOCATION_ENTITY_TYPE, tripDayEntity.getKey());
+    DomeEntity.setProperty(TripDay.NAME, DOME_ADDRESS);
+    DomeEntity.setProperty(TripDay.ORDER, 1);
+    datastore.put(DomeEntity);
+
+    Entity YosemiteEntity = new Entity(TripDay.LOCATION_ENTITY_TYPE, tripDayEntity.getKey());
+    YosemiteEntity.setProperty(TripDay.NAME, YOSEMITE_ADDRESS);
+    YosemiteEntity.setProperty(TripDay.ORDER, 0);
+    datastore.put(YosemiteEntity);
+
+    // run do Get
+    mapServlet.doGetMap(response, datastore, userEntity, tripEntity.getKey());
+
+    // even though DomeEntity is added first, its order property is 1 so it should appear 
+    // after YosemiteEntity in the JSON.
+    String expectedJson = "[\"" + INPUT_DESTINATION + "\",\""+YOSEMITE_ADDRESS+"\",\""+DOME_ADDRESS+"\"]";
+    
+    writer.flush();
+    Assert.assertTrue(stringWriter.toString().contains(expectedJson));
+  }
+
+  /* 
+   * Tests doGet when there is no current user
+   */
+  @Test
+  public void testNoCurrentUser() throws Exception {
+    HttpServletRequest request = mock(HttpServletRequest.class);       
+    HttpServletResponse response = mock(HttpServletResponse.class);  
+
+    when(request.getParameter("tripKey")).thenReturn("true");
+
+    // Create writers to check against actual output.
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    mapServlet.doGet(request, response);
+
+    String expectedJson = "No current User";
+    Assert.assertTrue(stringWriter.toString().contains(expectedJson));
+  }
+
+  /* 
+   * Tests doGet when there is no trip key
+   */
+  @Test
+  public void testNoTripKeyPassedIn() throws Exception {
+    HttpServletRequest request = mock(HttpServletRequest.class);       
+    HttpServletResponse response = mock(HttpServletResponse.class); 
+
+    // Create writers to check against actual output.
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    // initialize datastore
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // create user entity
+    Entity userEntity = new Entity(User.USER);
+    userEntity.setProperty(User.USER_EMAIL, EMAIL);
+    datastore.put(userEntity);
+
+    mapServlet.doGet(request, response);
+
+    String expectedJson = "No trip Key";
+    Assert.assertTrue(stringWriter.toString().contains(expectedJson));
```

Comment:
Do you need to create `userEntity`? It doesn't reach that in the `doGet` method. You may also want to mock the `getParameter("tripKey")` function to return `null`, just to show what is expected.

Reply:
True, but I just wanted to ensure that this was testing the case where no tripKey was provided, not the case where there is no user. I added the mock as you suggested!

Reply:
Ok, but just so you know, simply adding the `userEntity` to Datastore does not actually mean that the `doGet(...)` thinks the user is logged in. I would advocate for removing that, as it shouldn't matter whether or not that `userEntity` is in Datastore.

Reply:
Ah ok that makes sense! Removed :)